### ISSUE TITLE
Add a pre-commit hook for `ruff`

### ISSUE THREAD

**Issue by alexhroom** [RascalSoftware/python-RAT#42]
With `ruff` added as a linter, I would suggest adding a [pre-commit hook](https://pre-commit.com) configuration to the repository. Pre-commit is a piece of software which lets you hook code into the `git commit` command before the changes are committed (e.g. with a `ruff check` pre-commit hook, when you make a commit locally it runs `ruff check` and only proceeds with the commit if it passes). A pre-commit hook that just runs `ruff check --fix` and `ruff format` on a commit would avoid the time and CI carbon wasted in the all-too-common scenario of "agh, I just pushed to github but forgot to run the linter/formatter!"

**Comment by DrPaulSharp** [comment]
I have this set up in my local repository making use of ruff-pre-commit: https://github.com/astral-sh/ruff-pre-commit

We also have a github action that runs ruff (checking linting and formatting without making any changes) when code is pushed to the main branch or a pull request is opened: https://github.com/RascalSoftware/python-RAT/actions/workflows/run_ruff.yml

So, going forward we should consider whether to:

1) Include the pre-commit hook in the repo so it's standard for all users
2) Rely instead on the github actions for this - one possibility though is to set up the actions such that the tests do not start until the ruff action passes.

What do you think @StephenNneji?

**Comment by StephenNneji** [comment]
> Include the pre-commit hook in the repo so it's standard for all users

No, I believe the developer should be responsible for their local workflow. We can add instructions to the contributing.md on how to setup the pre-commit hook for anyone interested.  

> Rely instead on the github actions for this - one possibility though is to set up the actions such that the tests do not start until the ruff action passes.

Yes we should do this

**Comment by DrPaulSharp** [comment]
Yeah, that sounds good to me. @alexhroom?

**Comment by alexhroom** [comment]
sounds good!

**Comment by DrPaulSharp** [comment]
Closed by #59
